src/api/oauth2.rs

```diff
@@ -0,0 +1,293 @@
+        .ok_or(OAuthError::AuthCodeListenerTerminated)?;
+    let mut reader = BufReader::new(&stream);
+    let mut request_line = String::new();
+    reader
+        .read_line(&mut request_line)
+        .map_err(|_| OAuthError::AuthCodeListenerRead)?;
+
+    let redirect_url = request_line
+        .split_whitespace()
+        .nth(1)
+        .ok_or(OAuthError::AuthCodeListenerParse)?;
+    let code = get_code(&("http://localhost".to_string() + redirect_url));
+
+    let message = "Go back to your terminal :)";
+    let response = format!(
+        "HTTP/1.1 200 OK\r\ncontent-length: {}\r\n\r\n{}",
+        message.len(),
+        message
+    );
+    stream
+        .write_all(response.as_bytes())
+        .map_err(|_| OAuthError::AuthCodeListenerWrite)?;
+
+    code
+}
+
+// If the specified `redirect_uri` is HTTP, loopback, and contains a port,
+// then the corresponding socket address is returned.
+fn get_socket_address(redirect_uri: &str) -> Option<SocketAddr> {
+    let url = match Url::parse(redirect_uri) {
+        Ok(u) if u.scheme() == "http" && u.port().is_some() => u,
+        _ => return None,
+    };
+    let socket_addr = match url.socket_addrs(|| None) {
+        Ok(mut addrs) => addrs.pop(),
+        _ => None,
+    };
+    if let Some(s) = socket_addr {
+        if s.ip().is_loopback() {
+            return socket_addr;
+        }
+    }
+    None
+}
+
+/// Obtain a Spotify access token using the authorization code with PKCE OAuth flow.
+/// The redirect_uri must match what is registered to the client ID.
+pub fn get_access_token(
+    client_id: &str,
+    redirect_uri: &str,
+    scopes: Vec<&str>,
+) -> Result<OAuthToken, OAuthError> {
+    let auth_url = AuthUrl::new("https://accounts.spotify.com/authorize".to_string())
+        .map_err(|_| OAuthError::InvalidSpotifyUri)?;
+    let token_url = TokenUrl::new("https://accounts.spotify.com/api/token".to_string())
+        .map_err(|_| OAuthError::InvalidSpotifyUri)?;
+    let redirect_url =
+        RedirectUrl::new(redirect_uri.to_string()).map_err(|e| OAuthError::InvalidRedirectUri {
+            uri: redirect_uri.to_string(),
+            e,
+        })?;
+    let client = BasicClient::new(
+        ClientId::new(client_id.to_string()),
+        None,
+        auth_url,
+        Some(token_url),
+    )
+    .set_redirect_uri(redirect_url);
+
+    let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+
+    // Generate the full authorization URL.
+    // Some of these scopes are unavailable for custom client IDs. Which?
+    let request_scopes: Vec<oauth2::Scope> = scopes
+        .clone()
+        .into_iter()
+        .map(|s| Scope::new(s.into()))
+        .collect();
+    let (auth_url, _) = client
+        .authorize_url(CsrfToken::new_random)
+        .add_scopes(request_scopes)
+        .set_pkce_challenge(pkce_challenge)
+        .url();
+
+    println!("Browse to: {}", auth_url);
+    if let Err(err) = open::that(auth_url.to_string()) {
```

Comment:
I'm not super clear on the benefits, but I'd be interested in doing that with portals/ASPHD at some later time!

Reply:
I don't know what portals?ASPHD is :see_no_evil: is this a browser window you can use in your app? I stuck with opening the default browser because it allowed me to use my password manager on the spotify loging dialog on my postmarketos device :smile:

Reply:
its a terrible (but lovely) name that i always mispell for this lib: https://github.com/bilelmoussaoui/ashpd